Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -43,42 +47,143 @@ void ScrollAnchoringController::invalidateAnchorElement()
+        // TODO: we need to think about position: absolute
+        // TODO: figure out how to get scrollable area for renderer to check if it is maintaining scroll anchor
+        if (renderer->style().overflowAnchor() == OverflowAnchor::None || renderer->isStickilyPositioned() || renderer->isFixedPositioned() || renderer->isPseudoElement() || renderer->isAnonymousBlock())
+            return CandidateExaminationResult::Exclude;
+        if (&element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element) || (renderer->isInline() && !renderer->isAtomicInlineLevelBox()))
+            return CandidateExaminationResult::Skip;
+        auto boxRect = renderer->absoluteBoundingBoxRect();
+        if (!boxRect.width() || !boxRect.height())
+            return CandidateExaminationResult::Skip;
+        if (layoutViewport.contains(boxRect))
+            return CandidateExaminationResult::Select;
+        if (layoutViewport.intersects(boxRect))
+            return CandidateExaminationResult::Descend;
+    }
+    return CandidateExaminationResult::Skip;
+}
 
-    // TODO: iterate from the current scroller not document root
-    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
-        if (auto renderer = element.renderBox()) {
-            if (renderer->style().overflowAnchor() == OverflowAnchor::None || &element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
-                continue;
-            auto boxRect = renderer->absoluteBoundingBoxRect();
-            if (layoutViewport.intersects(boxRect))
-                candidateAnchor = &element;
-            if (layoutViewport.contains(boxRect))
-                break;
+#if !LOG_DISABLED
+static TextStream& operator<<(TextStream& ts, CandidateExaminationResult result)
+{
+    switch (result) {
+    case CandidateExaminationResult::Exclude:
+        ts << "Exclude";
+        break;
+    case CandidateExaminationResult::Select:
+        ts << "Select";
+        break;
+    case CandidateExaminationResult::Descend:
+        ts << "Descend";
+        break;
+    case CandidateExaminationResult::Skip:
+        ts << "Skip";
+        break;
+    }
+    return ts;
+}
+#endif
+
+Element* ScrollAnchoringController::findAnchorElementRecursive(Element* element)
+{
+    if (!element)
+        return nullptr;
+
+    auto result = examineCandidate(*element);
+    LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::findAnchorElementRecursive() element: "<< *element<<" examination result: " << result);
+
+    switch (result) {
+    case CandidateExaminationResult::Select:
+        return element;
+    case CandidateExaminationResult::Exclude:
+        return nullptr;
+    case CandidateExaminationResult::Skip:
+    case CandidateExaminationResult::Descend: {
+        for (auto& child : childrenOfType<Element>(*element)) {
+            if (auto* anchorElement = findAnchorElementRecursive(&child))
+                return anchorElement;
         }
     }
```

Comment:
Add a break at the end of this case statement, so if we add new ones in future we don't get fall-through by mistake.

Reply:
Fixed.